sage/gen_exhaustive_groups.sage

```diff
@@ -0,0 +1,83 @@
+# Define field size and field
+P = 2^256 - 2^32 - 977
+F = GF(P)
+
+orders_done = set()
+first = True
+for b in range(1, P):
+    E = EllipticCurve(F, [0, b])
+    n = E.order()
+    # Skip curves isomorphic to the real secp256k1
+    if n.is_pseudoprime():
+        continue
+    # Skip curves with an order we've already tried
+    if n in orders_done:
+        continue
+    orders_done.add(n)
+
+    # Find what prime subgroups exist
+    for f, _ in n.factor():
+        # Skip subgroups of order >1000
+        if f > 1000:
+            continue
+        # Skip subgroups that are not 3n+1 (needed for endomorphism)
+        if f % 3 != 1:
+            continue
+
+        if first:
+            print("#  if EXHAUSTIVE_TEST_ORDER == %i" % f)
+            first = False
+        else:
+            print("#  elif EXHAUSTIVE_TEST_ORDER == %i" % f)
+
+        # Iterate over X coordinates until we find one that is on the curve, has order f,
+        # and for which curve isomorphism exists that maps it to X coordinate 1.
+        for x in range(1, P):
+            # Skip X coordinates not on the curve, and construct the full point otherwise.
+            if not E.is_x_coord(x):
+                continue
+            G = E.lift_x(F(x))
+
+            # Skip points whose order is not a multiple of f. Project the point to have
+            # order f otherwise.
+            if (G.order() % f):
+                continue
+            G = G * (G.order() // f)
+            assert G.order() == f
+
+            # Now look for an isomorphism of the curve that gives this generator X
+            # coordinate equal to 1.
+            # If (x,y) is on y^2 = x^3 + b, then (a^2*x, a^3*y) is on y^2 = x^3 + a^6*b.
+            # So look for m=a^2=1/x.
+            m = F(1)/G[0]
+            if not m.is_square():
+                continue
+            rb = b*m^3
+            RE = EllipticCurve(F, [0, rb])
+
+            # Use as generator twice the point with this low X coordinate (like secp256k1!)
+            RG = RE.lift_x(1) * 2
```

Comment:
Why does `RG` have the same order as `G`?

Reply:
It's the image of G under the isomorphism described in the comments. I can make it more explicit by computing it directly instead of re-lifting from 1.

Reply:
Done.

Reply:
Oh I see now. You're specifically checking that `m` is square, i.e., that `a` exists.

Reply:
The new script output makes running this quite exciting!